File: backbone-go/application/exceptions/application_exceptions.go
// Package exceptions provides application layer exceptions with 8-digit error codes (10xxxxxx)
package exceptions

import (
	"fmt"
	"time"

	domainExceptions "github.com/freakjazz/backbone-go/domain/exceptions"
)

// ApplicationException represents application layer exceptions (10xxxxxx)
type ApplicationException struct {
	*domainExceptions.BaseKernelException
}

// NewApplicationException creates a new application exception
func NewApplicationException(code int, message string, details map[string]interface{}) *ApplicationException {
	if code < 10000000 || code > 10999999 {
		panic(fmt.Sprintf("Application exception code must be in range 10000000-10999999, got: %d", code))
	}

	return &ApplicationException{
		BaseKernelException: &domainExceptions.BaseKernelException{
			Code:        code,
			Message:     message,
			HTTPCode:    500,
			Details:     details,
			Layer:       "application",
			Severity:    domainExceptions.SeverityMedium,
			Timestamp:   time.Now().UTC(),
			Recoverable: true,
			Context:     make(map[string]interface{}),
		},
	}
}

// UseCaseException for use case failures
type UseCaseException struct {
	*ApplicationException
	UseCaseName string
}

// NewUseCaseException creates a use case exception
func NewUseCaseException(message, useCaseName string) *UseCaseException {
	details := map[string]interface{}{
		"use_case": useCaseName,
	}

	exception := NewApplicationException(10001001, message, details)

	return &UseCaseException{
		ApplicationException: exception,
		UseCaseName:          useCaseName,
	}
}

// ValidationError represents a single validation error
type ValidationError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
	Code    string `json:"code,omitempty"`
}

// ValidationException for validation errors
type ValidationException struct {
	*ApplicationException
	ValidationErrors []ValidationError
}

// NewValidationException creates a validation exception
func NewValidationException(message string, errors []ValidationError) *ValidationException {
	details := map[string]interface{}{
		"validation_errors": errors,
	}

	exception := NewApplicationException(10002001, message, details)
	exception.HTTPCode = 400

	return &ValidationException{
		ApplicationException: exception,
		ValidationErrors:     errors,
	}
}

// AuthorizationException for authorization failures
type AuthorizationException struct {
	*ApplicationException
	RequiredPermission string
	UserID             string
}

// NewAuthorizationException creates an authorization exception
func NewAuthorizationException(message, requiredPermission, userID string) *AuthorizationException {
	details := map[string]interface{}{
		"required_permission": requiredPermission,
		"user_id":             userID,
	}

	exception := NewApplicationException(10003001, message, details)
	exception.HTTPCode = 403

	return &AuthorizationException{
		ApplicationException: exception,
		RequiredPermission:   requiredPermission,
		UserID:               userID,
	}
}

// ResourceNotFoundException for resource not found errors
type ResourceNotFoundException struct {
	*ApplicationException
	ResourceType string
	ResourceID   string
}

// NewResourceNotFoundException creates a resource not found exception
func NewResourceNotFoundException(message, resourceType, resourceID string) *ResourceNotFoundException {
	details := map[string]interface{}{
		"resource_type": resourceType,
		"resource_id":   resourceID,
	}

	exception := NewApplicationException(10004001, message, details)
	exception.HTTPCode = 404

	return &ResourceNotFoundException{
		ApplicationException: exception,
		ResourceType:         resourceType,
		ResourceID:           resourceID,
	}
}

// ExternalServiceException for external service failures
type ExternalServiceException struct {
	*ApplicationException
	ServiceName string
	Operation   string
}

// NewExternalServiceException creates an external service exception
func NewExternalServiceException(message, serviceName, operation string) *ExternalServiceException {
	details := map[string]interface{}{
		"service":   serviceName,
		"operation": operation,
	}

	exception := NewApplicationException(10005001, message, details)
	exception.HTTPCode = 502

	return &ExternalServiceException{
		ApplicationException: exception,
		ServiceName:          serviceName,
		Operation:            operation,
	}
}

// Application Error Codes
const (
	// Use Case Errors (10001XXX)
	UseCaseError = 10001001

	// Validation Errors (10002XXX)
	ValidationErrorCode  = 10002001
	RequiredFieldMissing = 10002002
	InvalidFieldFormat   = 10002003
	FieldOutOfRange      = 10002004

	// Authorization Errors (10003XXX)
	AuthorizationError      = 10003001
	InsufficientPermissions = 10003002
	RoleNotAuthorized       = 10003003

	// Resource Errors (10004XXX)
	ResourceNotFound = 10004001
	UserNotFound     = 10004002
	EntityNotFound   = 10004003

	// External Service Errors (10005XXX)
	ExternalServiceError     = 10005001
	PaymentGatewayError      = 10005002
	NotificationServiceError = 10005003
	GeocodingServiceError    = 10005004
)

File: backbone-python/application/exceptions/application_exceptions.py
"""
Application layer exceptions - Códigos 10XXXXXX
"""
from typing import Dict, Any, Optional, List
from .base_application_exception import ApplicationException


class UseCaseException(ApplicationException):
    """
    Excepción para casos de uso fallidos.
    
    Ejemplos:
    - Caso de uso de registro fallido
    - Caso de uso de actualización fallido
    """
    
    def __init__(
        self,
        message: str,
        use_case_name: str,
        code: int = 10001001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        details.update({"use_case": use_case_name})
        kwargs['details'] = details
        
        super().__init__(code, message, **kwargs)


class ValidationException(ApplicationException):
    """
    Excepción para errores de validación en aplicación.
    
    Diferentes a las validaciones de dominio.
    Estas son validaciones de casos de uso.
    """
    
    def __init__(
        self,
        message: str,
        validation_errors: List[Dict[str, str]] = None,
        code: int = 10002001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        if validation_errors:
            details.update({"validation_errors": validation_errors})
        kwargs['details'] = details
        
        super().__init__(code, message, http_code=400, **kwargs)


class AuthorizationException(ApplicationException):
    """
    Excepción para errores de autorización.
    
    Usuario autenticado pero sin permisos.
    """
    
    def __init__(
        self,
        message: str,
        required_permission: str = None,
        user_id: str = None,
        code: int = 10003001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        if required_permission:
            details.update({"required_permission": required_permission})
        if user_id:
            details.update({"user_id": user_id})
        kwargs['details'] = details
        
        super().__init__(code, message, http_code=403, **kwargs)


class ResourceNotFoundException(ApplicationException):
    """
    Excepción para recursos no encontrados.
    """
    
    def __init__(
        self,
        message: str,
        resource_type: str,
        resource_id: str,
        code: int = 10004001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        details.update({
            "resource_type": resource_type,
            "resource_id": resource_id
        })
        kwargs['details'] = details
        
        super().__init__(code, message, http_code=404, **kwargs)


class ResourceConflictException(ApplicationException):
    """
    Excepción para conflictos de recursos.
    
    Ejemplos:
    - Email ya existe
    - Licencia ya registrada
    """
    
    def __init__(
        self,
        message: str,
        resource_type: str,
        conflict_field: str,
        conflict_value: str,
        code: int = 10005001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        details.update({
            "resource_type": resource_type,
            "conflict_field": conflict_field,
            "conflict_value": conflict_value
        })
        kwargs['details'] = details
        
        super().__init__(code, message, http_code=409, **kwargs)


class ApplicationServiceException(ApplicationException):
    """
    Excepción para servicios de aplicación.
    """
    
    def __init__(
        self,
        message: str,
        service_name: str,
        operation: str,
        code: int = 10006001,
        **kwargs
    ):
        details = kwargs.get('details', {})
        details.update({
            "service_name": service_name,
            "operation": operation
        })
        kwargs['details'] = details
        
        super().__init__(code, message, **kwargs)


# Catálogo de códigos de aplicación
class ApplicationErrorCodes:
    """Códigos de error específicos de aplicación 10XXXXXX"""
    
    # Use Cases (10001XXX)
    USE_CASE_ERROR = 10001001
    CREATE_USER_USE_CASE_ERROR = 10001002
    UPDATE_USER_USE_CASE_ERROR = 10001003
    DELETE_USER_USE_CASE_ERROR = 10001004
    CREATE_TRIP_USE_CASE_ERROR = 10001005
    
    # Validation (10002XXX)
    VALIDATION_ERROR = 10002001
    INPUT_VALIDATION_ERROR = 10002002
    BUSINESS_VALIDATION_ERROR = 10002003
    
    # Authorization (10003XXX)
    AUTHORIZATION_ERROR = 10003001
    INSUFFICIENT_PERMISSIONS = 10003002
    ROLE_REQUIRED = 10003003
    
    # Resource Not Found (10004XXX)
    RESOURCE_NOT_FOUND = 10004001
    USER_NOT_FOUND = 10004002
    DRIVER_NOT_FOUND = 10004003
    TRIP_NOT_FOUND = 10004004
    BOOKING_NOT_FOUND = 10004005
    
    # Resource Conflict (10005XXX)
    RESOURCE_CONFLICT = 10005001
    EMAIL_ALREADY_EXISTS = 10005002
    IDENTIFICATION_ALREADY_EXISTS = 10005003
    LICENSE_ALREADY_EXISTS = 10005004
    VEHICLE_PLATE_ALREADY_EXISTS = 10005005
    
    # Application Services (10006XXX)
    APPLICATION_SERVICE_ERROR = 10006001
    USER_SERVICE_ERROR = 10006002
    TRIP_SERVICE_ERROR = 10006003
    NOTIFICATION_SERVICE_ERROR = 10006004